### PR TITLE
[Merged by Bors] - Openshift-compatibility changes

### PR DESCRIPTION
**Pull request by adwk67**
# Description

Adds change to the kafka-operator so that it runs on OpenShift.
~N.B. TODO: change branch ref to a tag for `operator-rs` when the necessary changes have been released.~ Done

Selected tests on openshift 4.11:
image

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
# Author
- [x] Changes are OpenShift compatible
- [x] Helm chart can be installed and deployed operator works
- [x] Integration tests passed (for non trivial changes): all tests pass, albeit not all once!
```

```[tasklist]
# Reviewer
- [x] Code contains useful comments
- [x] (Integration-)Test cases added
- [x] Documentation added or updated
- [x] Changelog updated
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
```

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: rust/operator/src/kafka_controller.rs [hunk 1]
```diff
@@ -894,12 +924,14 @@ fn build_broker_rolegroup_statefulset(
...
-            run_as_group: Some(1000),
-            fs_group: Some(1000),
-            ..PodSecurityContext::default()
-        });
+        .service_account_name(sa_name)
+        .security_context(
+            PodSecurityContextBuilder::new()
+                .run_as_user(KAFKA_UID)
+                .run_as_group(0)
+                .fs_group(1000) // Needed for secret-operator
```
Comment (sbernauer) [inline rust/operator/src/kafka_controller.rs:932]: There are other reasons as well
```suggestion
                .fs_group(1000)
```

File: rust/operator/src/kafka_controller.rs [hunk 2]
```diff
@@ -894,12 +924,14 @@ fn build_broker_rolegroup_statefulset(
...
-            run_as_user: Some(1000),
-            run_as_group: Some(1000),
-            fs_group: Some(1000),
-            ..PodSecurityContext::default()
-        });
+        .service_account_name(sa_name)
+        .security_context(
+            PodSecurityContextBuilder::new()
+                .run_as_user(KAFKA_UID)
+                .run_as_group(0)
```
Comment (sbernauer) [inline rust/operator/src/kafka_controller.rs:931]: Is there any way we can relax this?

File: rust/crd/Cargo.toml [hunk 3]
```diff
@@ -9,7 +9,7 @@ version = "0.0.0-dev"
 publish = false
 
 [dependencies]
-stackable-operator = { git = "https://github.com/stackabletech/operator-rs.git", tag = "0.36.0" }
+stackable-operator = { git = "https://github.com/stackabletech/operator-rs.git", branch = "main" }
```
Comment (sbernauer) [inline rust/crd/Cargo.toml:12]: Marker to not forget about this

File: rust/operator/src/kafka_controller.rs (hunk 2 above)
Comment (adwk67) [reply rust/operator/src/kafka_controller.rs:931]: I don't think so: Openshift requires group:0 in order to dynamically assign UIDs.

File: rust/crd/Cargo.toml (hunk 3 above)
Comment (adwk67) [reply rust/crd/Cargo.toml]: Done.

File: rust/operator/src/kafka_controller.rs [hunk 4]
```diff
@@ -945,12 +919,7 @@ fn build_broker_rolegroup_statefulset(
     let pod_template_spec = pod_template.spec.get_or_insert_with(PodSpec::default);
     // Don't run kcat pod as PID 1, to ensure that default signal handlers apply
     pod_template_spec.share_process_namespace = Some(true);
-    pod_template_spec.service_account_name = Some(
-        serviceaccount
-            .name_in_ns_of(kafka)
-            .context(InvalidServiceAccountSnafu)?
-            .to_string(),
-    );
+    pod_template_spec.service_account_name = Some(sa_name.to_string());
```
Comment (maltesander) [inline rust/operator/src/kafka_controller.rs:922]: This is already set in line 899. Can be removed?

**Review by maltesander — CHANGES_REQUESTED** [review]
Needs some fixes. Starting tests now.

File: rust/crd/src/lib.rs [hunk 5]
```diff
@@ -12,10 +12,10 @@ use crate::tls::KafkaTls;
 use affinity::get_affinity;
 use serde::{Deserialize, Serialize};
 use snafu::{OptionExt, ResultExt, Snafu};
+use stackable_operator::commons::cluster_operation::ClusterOperation;
```
[the remark this answers is not in the document]
Comment (adwk67) [reply rust/crd/src/lib.rs]: Done.

File: rust/operator/src/kafka_controller.rs [hunk 6]
```diff
@@ -280,6 +300,22 @@ pub async fn reconcile_kafka(kafka: Arc<KafkaCluster>, ctx: Arc<Ctx>) -> Result<
...
+
+    let rbac_sa = cluster_resources
+        .add(client, rbac_sa)
+        .await
+        .context(ApplyServiceAccountSnafu)?;
+    cluster_resources
+        .add(client, rbac_rolebinding)
+        .await
+        .context(ApplyRoleBindingSnafu)?;
+
```
[the remark this answers is not in the document]
Comment (adwk67) [reply rust/operator/src/kafka_controller.rs]: Done.

**Comment by adwk67** [comment]
bors merge

**Pull request closed without merging**